File: WorldFeed/Server/src/Services/Identity/WorldFeed.Identity.API/Client/Validators/AuthClientParametersValidator.cs
namespace WorldFeed.Identity.API.Client.Validators
{
    using System;

    using WorldFeed.Common.Public.Parameters.Auth;
    using WorldFeed.Identity.API.Application.Requesters;

    public interface IAuthClientParametersValidator
    {
        void Validate(ICreateBearerTokenParameters parameters, ITwitterRequest request);

        void Validate(IRequestAuthUrlParameters parameters);

        void Validate(IRequestCredentialsParameters parameters);

        void Validate(IInvalidateAccessTokenParameters parameters, ITwitterRequest request);

        void Validate(IInvalidateBearerTokenParameters parameters, ITwitterRequest request);
    }

    public class AuthClientParametersValidator : IAuthClientParametersValidator
    {
        private readonly IAuthClientRequiredParametersValidator authClientRequiredParametersValidator;

        public AuthClientParametersValidator(IAuthClientRequiredParametersValidator authClientRequiredParametersValidator)
        {
            this.authClientRequiredParametersValidator = authClientRequiredParametersValidator;
        }

        public void Validate(ICreateBearerTokenParameters parameters, ITwitterRequest request)
        {
            this.authClientRequiredParametersValidator.Validate(parameters, request);
        }

        public void Validate(IRequestAuthUrlParameters parameters)
        {
            this.authClientRequiredParametersValidator.Validate(parameters);

            if (parameters.CallbackUrl != null)
            {
                if (parameters.CallbackUrl != "oob" && Uri.IsWellFormedUriString(parameters.CallbackUrl, UriKind.Absolute) == false)
                {
                    throw new ArgumentException("Invalid format, must be absolute uri or have a value of 'oob'", $"{nameof(parameters)}{nameof(parameters.CallbackUrl)}");
                }
            }
        }

        public void Validate(IRequestCredentialsParameters parameters)
        {
            this.authClientRequiredParametersValidator.Validate(parameters);
        }

        public void Validate(IInvalidateBearerTokenParameters parameters, ITwitterRequest request)
        {
            this.authClientRequiredParametersValidator.Validate(parameters, request);
        }

        public void Validate(IInvalidateAccessTokenParameters parameters, ITwitterRequest request)
        {
            this.authClientRequiredParametersValidator.Validate(parameters, request);
        }
    }
}

File: Sharebook/Client/src/app/core/Core/Client/Validators/AuthClientParametersValidator.ts
import {inject, Inject, Injectable, InjectionToken} from "@angular/core";

import {ITwitterRequest} from "../../../Public/Models/Interfaces/ITwitterRequest";
import {ICreateBearerTokenParameters} from "../../../Public/Parameters/Auth/CreateBearerTokenParameters";
import {IRequestAuthUrlParameters} from "../../../Public/Parameters/Auth/IRequestAuthUrlParameters";
import {IRequestCredentialsParameters} from "../../../Public/Parameters/Auth/RequestCredentialsParameters";
import {IInvalidateBearerTokenParameters} from "../../../Public/Parameters/Auth/InvalidateBearerTokenParameters";
import {IInvalidateAccessTokenParameters} from "../../../Public/Parameters/Auth/InvalidateAccessTokenParameters";
import {AuthParameters} from "./parameters-types";
import {UriKind} from "../../../Public/Models/Enum/uri-kind";
import {
  AuthClientRequiredParametersValidator,
  IAuthClientRequiredParametersValidator,
  IAuthClientRequiredParametersValidatorToken
} from "./AuthClientRequiredParametersValidator";
import ArgumentException from "typescript-dotnet-commonjs/System/Exceptions/ArgumentException";
import {UriExtensions} from "../../Extensions/UriExtensions";

export interface IAuthClientParametersValidator {
  validate(parameters: ICreateBearerTokenParameters, request: ITwitterRequest): void;

  validate(parameters: IRequestAuthUrlParameters): void;

  validate(parameters: IRequestCredentialsParameters): void;

  validate(parameters: IInvalidateAccessTokenParameters, request: ITwitterRequest): void;

  validate(parameters: IInvalidateBearerTokenParameters, request: ITwitterRequest): void;
}

export const IAuthClientParametersValidatorToken = new InjectionToken<IAuthClientParametersValidator>('IAuthClientParametersValidator', {
  providedIn: 'root',
  factory: () => new AuthClientParametersValidator(inject(AuthClientRequiredParametersValidator)),
});

@Injectable({
  providedIn: 'root',
})
export class AuthClientParametersValidator implements IAuthClientParametersValidator {
  private readonly _authClientRequiredParametersValidator: IAuthClientRequiredParametersValidator;

  constructor(@Inject(IAuthClientRequiredParametersValidatorToken) authClientRequiredParametersValidator: IAuthClientRequiredParametersValidator) {
    this._authClientRequiredParametersValidator = authClientRequiredParametersValidator;
  }

  public validate(parameters: AuthParameters, request?: ITwitterRequest): void {
    this._authClientRequiredParametersValidator.validate(parameters, request);

    if (AuthClientParametersValidator.isIRequestAuthUrlParameters(parameters)) {
      if (parameters.callbackUrl != null) {
        if (parameters.callbackUrl !== "oob" && !UriExtensions.isWellFormedUriString(parameters.callbackUrl, UriKind.Absolute)) {
          throw new ArgumentException("Invalid format, must be absolute uri or have a value of 'oob'", `${`nameof(parameters)`}${`nameof(parameters.callbackUrl)`}`);
        }
      }
    }
  }

  private static isIRequestAuthUrlParameters(parameters: AuthParameters): parameters is IRequestAuthUrlParameters {
    return (parameters as IRequestAuthUrlParameters).authAccessType !== undefined;
  }
}
